iMIP email improvements (take 2)

brad2014 opened the pull request:
This PR is a replacement for PR #17195. It is intended to be simpler
to review and approve, with fewer changes, some disabled by default.

fix #12391

It addresses issues #12391 and #13555, with the following changes:

- The plainText of iMIP emails has been upgraded as described in
issue #12391. The HTML design style has not been changed.

- Some of the HTML and plainText content has been rearranged
(simplified header language, moving the event title to from text
body to the first item in the bullet list, spelling corrections,
moving the description to the end of the list), per issue #12391.

- The interface for EMailTemplate has been extended: addBodyListItem
now takes an optional `plainIndent` parameter. Existing callers
see no change. Where new calls set the  new parameter >0, the list
item label (metaInfo) is put in column 1, and the value is indented
into column 2 (properly accounting for multiple lines, if any).

- An optional dav config setting has been added,
`invitation_list_attendees`. It defaults to 'no', leaving emails
unchanged. If set by the site admin to 'yes', then iMIP emails
include, for the organizer and each attendee, their name, email,
and a ✔︎ if they have accepted the invitation.

- Minor refactoring.

Notes:

- The labels for organizers and attendees list items are new, and
require translation/localization.

- Dav config settings are documented in the code, but not in the
Administrator's Guide.

- There is no administrative UI to change the "include attendees" setting.  Only via occ.

Signed-off-by: Brad Rubenstein brad@wbr.tech

brad2014 committed:
iMIP email improvements (take 2)

This PR is a replacement for PR #17195. It is intended to be simpler
to review and approve, with fewer changes, some disabled by default.

It addresses issues #12391 and #13555, with the following changes:

- The plainText of iMIP emails has been upgraded as described in
issue #12391. The HTML design style has not been changed.

- Some of the HTML and plainText content has been rearranged
(simplified header language, moving the event title to from text
body to the first item in the bullet list, spelling corrections,
moving the description to the end of the list), per issue #12391.

- The interface for EMailTemplate has been extended: addBodyListItem
now takes an optional `plainIndent` parameter. Existing callers
see no change. Where new calls set the  new parameter >0, the list
item label (metaInfo) is put in column 1, and the value is indented
into column 2 (properly accounting for multiple lines, if any).

- An optional dav config setting has been added,
`invitation_list_attendees`. It defaults to 'no', leaving emails
unchanged. If set by the site admin to 'yes', then iMIP emails
include, for the organizer and each attendee, their name, email,
and a ✔︎ if they have accepted the invitation.

- Minor refactoring.

Notes:

- The labels for organizers and attendees list items are new, and
require translation/localization.

- Dav config settings are documented in the code, but not in the
Administrator's Guide.

Signed-off-by: brad2014 <brad2014@users.noreply.github.com>

brad2014 commented:
Here is a screenshot (from thunderbird) of the resulting imip mail as plain text:  

<img width="605" alt="new text" src="https://user-images.githubusercontent.com/6502462/66376496-aa058280-e964-11e9-96a4-d580a5ee64c3.png">

and as html:

<img width="625" alt="new html" src="https://user-images.githubusercontent.com/6502462/66376485-9fe38400-e964-11e9-9c70-1ba568679979.png">

brad2014 commented:
If we do 
```
# sudo -u apache php occ config:app:set dav invitation_list_attendees --value yes
```
Then the plain text looks like this (notice that "brad 2" is seeing that "brad 1" has accepted):

<img width="599" alt="new text with attendees" src="https://user-images.githubusercontent.com/6502462/66376633-f6e95900-e964-11e9-94d4-0a603327effa.png">

and the html looks like this:

<img width="648" alt="new html with attendees" src="https://user-images.githubusercontent.com/6502462/66376673-07013880-e965-11e9-859e-b2390590dc00.png">

brad2014 commented:
@skjnldsv and @nextcloud/designers

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -95,6 +95,7 @@ class IMipPlugin extends SabreIMipPlugin {
 	const METHOD_REQUEST = 'request';
 	const METHOD_REPLY = 'reply';
 	const METHOD_CANCEL = 'cancel';
+	const IMIP_INDENT = 15; // Enough for the length of all body bullet items, in all languages
```
This activates the new plainText formatting in EmailTemplate::addBodyListItem.

Was, for example:  
```
  * My house (Where:)
  * These are multi-line
event details! (Description:)
```
Is now, iff plainIndent>0:
```
         Where: My house
   Description: These are
                event details!

Calculating a reasonable indent dynamically is TODO.  In the mean time, this degrades gracefully if, in some language, a label is too long (it just means the value will be misaligned, but still readable).  Note that this doesn't handle right-to-left languages *at all* (another TODO).

skjnldsv replied:
What if the label is too wide? What is done currently? Ellipsised?

brad2014 replied:
> What if the label is too wide? What is done currently? Ellipsised?

No, it would just disturb the alignment:
```
         Titel: xxxxxxxx
          Zeit: xxxxxxxxxxxxx
Straßenbahnhaltestelle: xxxxxxxxxxxx
          Link: xxxxxxxxxxxx
```

skjnldsv replied:
Can we ellipsis instead? :)

brad2014 replied:
> Can we ellipsis instead? :)

I'd advise against it. Given the choice between readability and alignment, I'd choose readability.  It is quite hard, in general, across languages, to determine the printed length of a unicode plainText string (you can't simply count up the bytes), so programmatically cutting off a label without context would reduce readability without even necessarily getting the alignment fixed.

A better solution would be to advise the translator that these labels sit in a limited width, and to pick a translation that is worded, abbreviated, or elided in a way that makes it fit (Practically speaking, I suspect this will almost never be an issue). 

```
         Titel: xxxxxxxx
          Zeit: xxxxxxxxxxxxx
 SBHaltestelle: xxxxxxxxxxxx
          Link: xxxxxxxxxxxx

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -180,26 +181,6 @@ public function schedule(Message $iTipMessage) {
-			$end = clone $vevent->DTSTART;
-		}
-
-		$meetingWhen = $this->generateWhenString($l10n, $start, $end);
```
Moved into the generateWhenString function, where it belongs.

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -478,49 +476,127 @@ private function isDayEqual(\DateTime $dtStart, \DateTime $dtEnd) {
+	 */
+
+	private function addAttendees(IEMailTemplate $template, IL10N $l10n, VEvent $vevent) {
+		if ($this->config->getAppValue('dav', 'invitation_list_attendees', 'no') === 'no') {
```
This new function is a no-op until the admin enables it (it is disabled by default, to match current e-mails, and because there may be policy reasons why iMip mails should not disclose so prominently the names and email addresses of all invited attendees (though there is nothing in the text that is not also in the ics file).

It adds to the list (before the description) the name/email/accept-status of the organizer and each attendee.

TODO: Really, the checkmark for "accepted" is lovely, but too limited.  There should be a mark for (a) not replied yet, (b) accepted, (c) declined.

brad2014 commented on lib/private/Mail/EMailTemplate.php:
```diff
@@ -454,14 +454,16 @@ public function addBodyText(string $text, $plainText = '') {
 	 */
-	public function addBodyListItem(string $text, string $metaInfo = '', string $icon = '', $plainText = '', $plainMetaInfo = '') {
+	public function addBodyListItem(string $text, string $metaInfo = '', string $icon = '', $plainText = '', $plainMetaInfo = '', $plainIndent = 0) {
 		$this->ensureBodyListOpened();
```
This is a (backward-compatible) interface change to the EMailTemplate interface.

If the plainIndent is not specified by the caller, this function is unchanged.  As of this commit, iMipPlugin.php is the only caller that uses it.  Other similar emails (event notifications, for example), might use it in the future.

brad2014 commented on lib/private/Mail/EMailTemplate.php:
```diff
@@ -454,14 +454,16 @@ public function addBodyText(string $text, $plainText = '') {
 		if ($plainText === '') {
 			$plainText = $text;
 			$text = htmlspecialchars($text);
+			$text = str_replace("\n", "<br/>", $text); // convert newlines to HTML breaks
```
This fixes an oops.   Newlines in plaintext are line-breaks in HTML.

brad2014 commented on lib/private/Mail/EMailTemplate.php:
```diff
@@ -542,7 +560,7 @@ public function addBodyButtonGroup(string $textLeft,
 
 		$this->htmlBody .= vsprintf($this->buttonGroup, [$color, $color, $urlLeft, $color, $textColor, $textColor, $textLeft, $urlRight, $textRight]);
-		$this->plainBody .= $plainTextLeft . ': ' . $urlLeft . PHP_EOL;
+		$this->plainBody .= PHP_EOL . $plainTextLeft . ': ' . $urlLeft . PHP_EOL;
```
This change mimics, for the plainText version of the e-mail, the fact that the buttons have space above them, separating them from the List.  It looks better.

skjnldsv reviewed (commented):
Looks really great!

On the first html screenshot, the organizer is a "contacts" icon, can we have multiple organizers in an event?
If not it should be a single-user icon :)

> * There is no administrative UI to change the "include attendees" setting.  Only via occ.

This is not acceptable, please create a config.php key or a config toggle in the mail settings! ;)


@georgehrke this raises a question to me, why do we have this mails in server and not in calendar?

tcitworld commented:
> @georgehrke this raises a question to me, why do we have this mails in server and not in calendar?

> Beginning with Nextcloud 9 the CalDAV server became part of Nextcloud's server and the calendar app is just a CalDAV client running in your browser.

georgehrke commented:
> On the first html screenshot, the organizer is a "contacts" icon, can we have multiple organizers in an event?

It can only be one. Ideally we would show the user avatar, but an icon is a good start for now. :) 

> why do we have this mails in server and not in calendar?

What @tcitworld said.
This email is sent out for all invitations, no matter whether you create them on your Android, your desktop client or in the calendar app.

skjnldsv commented:
Ah right, I forgot! :)

brad2014 commented:
> This is not acceptable, please create a config.php key or a config toggle in the mail settings! ;)

That will be in the very next PR, I promise. ;)

brad2014 commented:
> > On the first html screenshot, the organizer is a "contacts" icon, can we have multiple organizers in an event?
> 
> It can only be one. Ideally we would show the user avatar, but an icon is a good start for now. :)

I limited myself to the svg's that are currently shipped in core/img/filetypes, which is where the icons were pulled prior.  The two icons that are "not quite right" are the Organizer (a single person) and the Title (which uses the text.svg icon, same as "Description"). It seems whoever chose the original design was being opportunistic/efficient in reusing the icons available for filetypes, and I didn't want to disturb that.  But if there is a currently shipping 16x16 svg icon representing a single user, let me know and I'll put it in.

These icons are, semantically, bullets, very small (16x16) and marking the separate paragraphs, so it wouldn't work to use the organizer's avatar photo without changing the design of the email (which I tried last time in PR #17195, removing the icons entirely, and that did not prove too popular).

For this PR, I'm trying to keep it super simple, on the assumption that a more competent person later will do a proper design refresh of the email layout.

skjnldsv commented:
> I limited myself to the svg's that are currently shipped in core/img/filetypes, which is where the icons were pulled prior. The two icons that are "not quite right" are the Organizer (a single person) and the Title (which uses the text.svg icon, same as "Description"). It seems whoever chose the original design was being opportunistic/efficient in reusing the icons available for filetypes, and I didn't want to disturb that. But if there is a currently shipping 16x16 svg icon representing a single user, let me know and I'll put it in.

yes there is :)
https://docs.nextcloud.com/server/stable/developer_manual/design/icons.html

The `user` icon in the actions folder

brad2014 committed:
Make icon set for CalDav event presentation, use in iMIP email.

iMIP e-mails were, as a hack, using filetypes icons for caldav
elements (titles, locations).  This commit creates a folder of
caldav element icons. To start, they are used in iMIP emails,
but eventually should be used by any app that wants to have icon
labels for caldav elements.

Signed-off-by: brad2014 <brad2014@users.noreply.github.com>

brad2014 commented:
Notes on 5dfbe56:

@skjnldsv, I took your suggestion and used the user.svg logo, though I had to recolor it.

@georgehrke, the icons used in iMip email were loaded from core/img/filetypes, as a hack (a caldav entry is not a filetype).  This commit creates the folder core/img/caldav, and it contains all the images for elements of an event.  These are used in iMIP email, but eventually should probably be used in any UI that refers to caldav elements (e.g. the calendar app). The images currently in the new folder are copied from the filetype images, except for "title" and "organizer" which were taken from core/img/actions and recolored to fit the existing email theme.

jancborchardt commented:
Cool stuff on improving that email! :) Some design feedback, cause the main content currently has lots of different elements so that everything gets lost a bit:

- The title of the event should be shown as the heading, where currently it only says "Invitation". Doesn’t need to be shown in the content then.
- Time/Date doesn’t need a "Time" heading. But the Date and Time should be shown on separate lines. Also the date should be written out, that is "Monday" and "October" instead of "Mon" and "Oct".
- "Location" also doesn’t need that heading, could ideally be a bit conversational instead simply using "At " before it on the same line. If that works with localization, if there’s languages which put that at the end.
- Organizer and Attendees can be combined. Organizers can simply be shown first in the attendee list, with "(organizer)" behind their name.
- We never use italics because it is not so ideal for accessibility. The headings like "Description" etc are greyish, that’s already enough.
- The Accept button should be on the right, Decline on the left. :)

What do you think? Might have more feedback after those changes are applied, but that’s it for now. Really nice that you work on improving this @brad2014!

brad2014 commented:
>     * Organizer and Attendees can be combined. Organizers can simply be shown first in the attendee list, with "(organizer)" behind their name.

The organizer is not necessarily an attendee, it is spec'd as an admin or point of contact. In many contexts the organizer is the person collecting the rsvp's, and managing the attendees list. The ical RFC has separate support for marking participation ROLE, e.g. of an attendee as "Chair" (the leader of a meeting), and that could be called out in the UI, but I think it's overkill (e.g. the calendar app UI won't let you set the participation role of each attendee).

skjnldsv commented:
Ping @jancborchardt, please check the answer above :)

brad2014 commented:
I'm still interested in making progress on this, if you are, @skjnldsv and @jancborchardt.  Was my reply helpful?

jancborchardt commented:
@brad2014 yes, sorry for the late reply! :) Sounds good to do enhancements in a follow-up pull request. (And yes, removing italics is fine – as said we do not use them anywhere as they are quite bad for readability.)

georgehrke commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -478,49 +476,127 @@ private function isDayEqual(\DateTime $dtStart, \DateTime $dtEnd) {
-			$template->addHeading($l10n->t('Invitation canceled'), $l10n->t('Hello %s,', [$attendeeName]));
-			$template->addBodyText($l10n->t('The meeting »%1$s« with %2$s was canceled.', [$summary, $inviteeName]));
+			$template->setSubject('Canceled: ' . $summary);
+			$template->addHeading($l10n->t('Invitation canceled'));
```
👍

georgehrke commented:
@brad2014 Can you please rebase this PR? :)

ChristophWurst commented:
I think due to the lack of activity and the fact that we'll also need follow-up fixes to fully have this done I'm going to move the milestone.

brad2014 commented:
> I think due to the lack of activity and the fact that we'll also need follow-up fixes to fully have this done I'm going to move the milestone.

Of course, I'm a bit disappointed, since it was my first pull request, and I redid it twice to get the scope to something that was reviewable and valuable on its own.  😅 

But I understand these things sometimes happen - It sounds like 19 was just too full.  There are no "follow-up fixes" that I know of, just ongoing incremental UX enhancements.  I'll set this aside until someone can tell me what (if any) are the minimal changes that need to happen to get it pushed to production, or that it's not in line with where y'all are going.

brad2014 commented:
> @brad2014 Can you please rebase this PR? :)

I tried to rebase, and I think I've completely borked everything up. I'm going to go into a corner and cry.

jancborchardt commented:
So we need a code review here, right? @ChristophWurst @georgehrke @gary-kim @StCyr @ma12-co want to check it out? :)

brad2014 commented:
> So we need a code review here, right? @ChristophWurst @georgehrke @gary-kim @StCyr @ma12-co want to check it out? 

Would anyone be willing to look at this?  Is there anything I can do to help move it along?

MorrisJobke commented:
> So we need a code review here, right? @ChristophWurst @georgehrke @gary-kim @StCyr @ma12-co want to check it out? :)

🏓 Would be nice to get this in.

georgehrke commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -506,49 +503,126 @@ private function isDayEqual(\DateTime $dtStart, \DateTime $dtEnd) {
+		}
+		if ($vevent->URL) {
+			$template->addBodyListItem(sprintf('<a href="%s">%s</a>',
+					htmlspecialchars($vevent->URL),
```
@nickvergessen Does this look sane?

nickvergessen replied:
looks good

georgehrke commented:
Posting some screenshots for easier review:

|Title | Screenshot |
|:---------:|:------:|
|Standard invite |![B28F041D-89BC-479A-BC49-981FD605E815](https://user-images.githubusercontent.com/1250540/90810859-f01cc800-e323-11ea-9a36-6b1be24bb3f6.png)|
|Attendee list|![0BC83560-9355-4BF6-A38D-0B4F13FD3BF6](https://user-images.githubusercontent.com/1250540/90810844-e98e5080-e323-11ea-8489-6046681dd74d.png)|
|Cancelation|![80126240-B156-497E-9DFA-171BE425B3EC](https://user-images.githubusercontent.com/1250540/90811018-2d815580-e324-11ea-9dbc-cc2b23672132.png)|

georgehrke commented:
The cancelation email does not include the entire attendee list, but I think that's related to Sabre/DAV not providing the entire list for `METHOD:CANCEL`. That's something we can fixup in a follow up PR.

georgehrke reviewed (approved):
👍 from my side.
~~Just waiting for @nickvergessen's feedback in https://github.com/nextcloud/server/pull/17456/files#r474167398~~

georgehrke commented:
@brad2014 I'm deeply sorry for taking so long :/

georgehrke commented:
@MorrisJobke What do we have to do to get CI results? Rebase? New PR?

MorrisJobke commented:
> @MorrisJobke What do we have to do to get CI results? Rebase? New PR?

Rebase I would say. Last push was in April - so master could be a lot different now.

georgehrke commented:
@brad2014 Could you please rebase? Thx!

brad2014 commented:
> The cancelation email does not include the entire attendee list, but I think that's related to Sabre/DAV not providing the entire list for `METHOD:CANCEL`. That's something we can fixup in a follow up PR.

Just a note: RFC 2446 indicates (via example) that when the organizer removes an individual from a meeting, the individual receives a METHOD:CANCEL event containing only their name in the attendee list.   If the meeting itself is canceled, all attendees receive a METHOD:CANCEL containing all the invited attendees (including those who have declined), along with STATUS:CANCELLED.

brad2014 commented:
> @brad2014 Could you please rebase? Thx!

Done.  There were no conflicts.

brad2014 commented:
> > @brad2014 Could you please rebase? Thx!
> 
> Done. There were no conflicts.

Hold off briefly on merge.  The static-code-analysis points to some possible errors that concern me.  Checking...
cc @MorrisJobke @georgehrke

brad2014 commented on lib/private/Mail/EMailTemplate.php:
```diff
@@ -447,19 +447,21 @@ public function addBodyText(string $text, $plainText = '') {
 		$this->ensureBodyListOpened();
 
-		if ($plainText === '') {
+		if ($plainText === '' || $plainText === true) {
 			$plainText = $text;
 			$text = htmlspecialchars($text);
+			$text = str_replace("\n", "<br/>", $text); // convert newlines to HTML breaks
```
The code handled the case where plainText and plainMetaInfo were strings or false booleans, but left the true bool case undefined.  This made the static-code-checker unhappy (which it should).  I simply handled the true case as if it were an empty string.

brad2014 commented:
> Hold off briefly on merge. The static-code-analysis points to some possible errors that concern me. Checking...
> cc @MorrisJobke @georgehrke

Commit 6a44945 gives clean Psalm output.  It fixes one bug (which would occur if a VEVENT had no DTEND, legal but rare), and the rest is type documenting.

I'm going to do some more testing tomorrow, to make sure I didn't break anything else.  I'd appreciate a second pair of 👀 if you have a moment.

georgehrke commented:
@brad2014 The PHP Codestyle check is failing. Can you please check.

georgehrke commented:
@nickvergessen @ChristophWurst @skjnldsv Can we please get a second review here? :)

brad2014 commented:
> @tcitworld: @brad2014 Edit `appinfo/info.xml` and set `max-version` to 20.

For the google:  occ app:enable has a -f flag to override version checking :-) 
```
# runuser -u apache -- php occ app:enable -f calendar
```

brad2014 commented:
@georgehrke 

> @brad2014 The PHP Codestyle check is failing. Can you please check.
> 
> You can just use `composer run cs:fix` to autofix it.

@georgehrke I followed the directions in 

> @brad2014 The PHP Codestyle check is failing. Can you please check.
> 
> You can just use `composer run cs:fix` to autofix it.

I installed composer, then I 

> @brad2014 The PHP Codestyle check is failing. Can you please check.
> 
> You can just use `composer run cs:fix` to autofix it.

@georgehrke I get this error.  Is the 3rdparty repo missing nextcloud\codingstandard?  Any pointers?
```
> php-cs-fixer fix
PHP Fatal error:  Uncaught Error: Class 'Nextcloud\CodingStandard\Config' not found in /var/www/nextcloud/.php_cs.dist:9
```

georgehrke commented:
Did you do a `composer i` to install all composer dependencies?

brad2014 commented:
> Did you do a `composer i` to install all composer dependencies?

I'll give it a go.  (Sorry if this is the wrong place for these newbie questions - I've never used composer before.)

brad2014 commented:
> > Did you do a `composer i` to install all composer dependencies?
> 
> I'll give it a go. (Sorry if this is the wrong place for these newbie questions - I've never used composer before.)

Thanks @georgehrke; I ran `composer i` per your suggestion, and it appears to create a lot of changes, which I don't think I should commit to my branch, since the reviews have been done, and I can't vouch for them.

```
$ composer i
... lots of messages, completes without error ...
$ git status
Refresh index: 100% (8883/8883), done.
On branch feature/brad2014/12391-improve-imip-mail-message-take-2
Your branch is up to date with 'origin/feature/brad2014/12391-improve-imip-mail-message-take-2'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   apps/accessibility/l10n/pt_BR.json
	modified:   apps/dav/lib/Connector/Sabre/ShareTypeList.php
	modified:   apps/federatedfilesharing/l10n/uk.js
	modified:   apps/files_external/3rdparty/icewind/smb/src/Exception/ConnectionException.php
	modified:   apps/oauth2/l10n/es_NI.js
	modified:   apps/settings/tests/Settings/Personal/Security/PasswordTest.php
	modified:   core/Migrations/Version13000Date20170926101637.php
	modified:   lib/composer/composer/LICENSE
	modified:   lib/composer/composer/autoload_classmap.php
	modified:   lib/composer/composer/autoload_psr4.php
	modified:   lib/composer/composer/autoload_real.php
	modified:   lib/composer/composer/autoload_static.php
	modified:   lib/composer/composer/installed.json
	modified:   lib/private/legacy/OC_App.php
	modified:   tests/lib/AllConfigTest.php

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	lib/composer/bin/
	lib/composer/composer/autoload_files.php
	lib/composer/composer/semver/
	lib/composer/composer/xdebug-handler/
	lib/composer/doctrine/
	lib/composer/friendsofphp/
	lib/composer/nextcloud/
	lib/composer/paragonie/
	lib/composer/php-cs-fixer/
	lib/composer/psalm/
	lib/composer/psr/
	lib/composer/symfony/

no changes added to commit (use "git add" and/or "git commit -a")
```

brad2014 commented:
> @brad2014 The PHP Codestyle check is failing. Can you please check.
> 
> You can just use `composer run cs:fix` to autofix it.

Done.

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -402,20 +400,19 @@ private function getAttendeeRSVP(Property $attendee = null) {
 			$isFloating = $vevent->DTSTART->isFloating();
 			$dtend = clone $vevent->DTSTART;
-			$endDateTime = $end->getDateTime();
+			$endDateTime = $dtend->getDateTime();
```
This was actually a bug, but would only happen if an event had no DTEND (legal per RFC, but rare in the field).

rullzer commented:
CI is not happy...

brad2014 commented:
> CI is not happy...



> CI is not happy...

I'm looking at it.

brad2014 commented:
For example:

```
# runuser -u apache -- NOCOVERAGE=true TEST_SELECTION=NODB ./autotest.sh sqlite
Using PHP executable /usr/bin/php
Using database oc_autotest
Setup environment for sqlite testing on local storage ...
Updated 0 paths from the index
Installing ....
Nextcloud was successfully installed
Testing with sqlite ...
No coverage
/var/www/nctest/lib/composer/bin/phpunit --configuration phpunit-autotest.xml --exclude-group DB,SLOWDB --log-junit autotest-results-sqlite.xml  
PHP Fatal error:  Uncaught Error: Call to a member function addPsr4() on bool in /var/www/nctest/lib/private/legacy/OC_App.php:277
Stack trace:
#0 /var/www/nctest/lib/private/AppFramework/Bootstrap/Coordinator.php(100): OC_App::registerAutoloading()
```

brad2014 commented:
[UPDATE] ha ha. It turns out `composer require --dev phpunit/phpunit 8.5.2` breaks everything, and installing phpunit-8.5.2.phar in /usr/local/bin is the way to go.

Commit:
Minor cleanup:  php-cs-fixer, tests, interface consistency

IMipPlugin.php Removed blank lines to make php-cs-fixer happy.
Minor cleanup:  bugs found by Psalm static checker
IEMailTemplate: The public interface to addBodyListItem also needs to include the new plainIndent parameter.
IMipPlugin: Fixes an undefined variable for events that do not have DTEND.  Also use explicit string conversion for parameters and properties in several places.

The new email template adds an additional blank line before "button" links in plain text, so the tests were fixed to include that additional blank line.

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

brad2014 commented:
There is an acceptance test failure in tests/lib/Avatar/GuestAvatarTest.php:70 that has nothing to do with my changes.
It looks like the binary representation of the guest avatar image has changed.

A cursory glance suggests that the image output of ImageMagick has changed slightly, and the resulting PNG data is visually the same, but internally slightly different.

But I don't know if you want me to update the test baseline data as part of this PR (accepting the new result), or let someone who knows that part of the code check it.

brad2014 commented:
CI tests for this PR now pass.

skjnldsv commented:
Whoop whoop! :rocket:
